# SSLv3 Support Patch for curl 7.77.0 with OpenSSL 3.x
# NOTE: Use this patch only for curl 7.77.0
# 
# This patch enables SSLv3 protocol support for security testing purposes.
# SSLv3 is an obsolete and insecure protocol (deprecated since 2015) but may
# be needed to verify that servers properly reject SSLv3 connections.
#
# OpenSSL 3.x removed the SSLv3_client_method() function, so this patch uses
# the modern SSL_CTX_set_min_proto_version() and SSL_CTX_set_max_proto_version()
# API instead, which is compatible with OpenSSL 3.x.
#
# Changes:
# 1. Remove "No SSLv3 support" error in method selection
# 2. Add CURL_SSLVERSION_SSLv3 case handler that sets protocol version to SSL3
# 3. Explicitly disable all other SSL/TLS protocol versions
#
# Usage: patch --ignore-whitespace -N lib/vtls/openssl.c sslv3.patch
#
--- openssl.c	2022-05-30 01:05:13.000000000 -0700
+++ openssl.c.2	2022-05-30 01:25:52.000000000 -0700
@@ -2709,8 +2709,9 @@
     failf(data, "No SSLv2 support");
     return CURLE_NOT_BUILT_IN;
   case CURL_SSLVERSION_SSLv3:
-    failf(data, "No SSLv3 support");
-    return CURLE_NOT_BUILT_IN;
+    req_method = SSLv3_client_method();
+    /* use_sni(FALSE); */
+    break;
   default:
     failf(data, "Unrecognized parameter passed via CURLOPT_SSLVERSION");
     return CURLE_SSL_CONNECT_ERROR;
@@ -2798,9 +2799,18 @@
 
   switch(ssl_version) {
     case CURL_SSLVERSION_SSLv2:
-    case CURL_SSLVERSION_SSLv3:
       return CURLE_NOT_BUILT_IN;
 
+    case CURL_SSLVERSION_SSLv3:
+      SSL_CTX_set_min_proto_version(backend->ctx, SSL3_VERSION);
+      SSL_CTX_set_max_proto_version(backend->ctx, SSL3_VERSION);
+      ctx_options |= SSL_OP_NO_SSLv2;
+      ctx_options |= SSL_OP_NO_TLSv1;
+      ctx_options |= SSL_OP_NO_TLSv1_1;
+      ctx_options |= SSL_OP_NO_TLSv1_2;
+      ctx_options |= SSL_OP_NO_TLSv1_3;
+      break;
+
     /* "--tlsv<x.y>" options mean TLS >= version <x.y> */
     case CURL_SSLVERSION_DEFAULT:
     case CURL_SSLVERSION_TLSv1: /* TLS >= version 1.0 */
